### PR TITLE
Add support for `I256` and `U256`

### LINKED ISSUE #1097: Add `AlgebraicType/Value::U256`
Many cryptographic hashes result in 32 byte (256 bit) outputs, and storing this as a `Box<U256>` is about as cheap as we can make it. This new variant/type would be helpful for e.g., `Identity`.

### PR DESCRIPTION
**Pull request by Centril**
# Description of Changes

Fixes #1097.

Implements support for `256`-width integers in SATS and friends.
The integer types are backed by the crate `ethnum` which provides support as-if they were primitive Rust types.

# API and ABI breaking changes

Breaks the BSATN-encoding of `AlgebraicType`.

# Expected complexity level and risk

2 -- some subtle considerations:
- sequences (supported, but can at most become `u128::MAX` for e.g., `u256`)

# Testing

Unit tests and SDK tests have been amended.

### CODE REVIEW

File: crates/bindings-csharp/BSATN.Runtime/BSATN/Int128.cs
Comment (RReverser) [inline crates/bindings-csharp/BSATN.Runtime/BSATN/Int128.cs:8]: This would be nice, but might be problematic if another package that user depends on declares its own `System.Int128` polyfill. It's ok to declare `internal` System types for things like marker attributes, but public types can lead to all sorts of conflicts.

I think we should keep this type under our own `SpacetimeDB` namespace.

File: crates/bindings-csharp/BSATN.Runtime/BSATN.Runtime.csproj
Comment (RReverser) [inline crates/bindings-csharp/BSATN.Runtime/BSATN.Runtime.csproj:11]: Good catch, this comment is outdated as of recently. It should be something like this though:

```suggestion
    <!-- Note: the binary produced by this package is used in Unity too, which is limited to .NET Standard 2.1. -->
```

File: crates/bindings-csharp/BSATN.Codegen/Utils.cs
Comment (RReverser) [inline crates/bindings-csharp/BSATN.Codegen/Utils.cs:136]: For custom types it's better to inline the `struct BSATN` implementation right inside the type instead of augmenting the codegen.

The only reason this codegen has a large switch is because C# doesn't have Rust-like traits that would allow us to declare BSATN implementations on built-in types - but where types are not built-in, we don't need to do this.

File: crates/bindings-csharp/BSATN.Runtime/BSATN/Int256.cs
Comment (RReverser) [inline crates/bindings-csharp/BSATN.Runtime/BSATN/Int256.cs:12]: Might want to add `IComparable<Int256>` here and to the other types.

File: crates/bindings-csharp/BSATN.Runtime/BSATN/Int256.cs
Comment (RReverser) [inline crates/bindings-csharp/BSATN.Runtime/BSATN/Int256.cs:132]: Do we want to provide proper ToString() maybe? Should be easy to do via BigInteger - not the most efficient method perhaps, but should be fine for stringification.

File: crates/bindings-csharp/BSATN.Runtime/BSATN/Int256.cs
Comment (RReverser) [inline crates/bindings-csharp/BSATN.Runtime/BSATN/Int256.cs:38]: I think these all implementations could be significantly simplified if you declared this type as a wrapper around `(UInt128 Upper, UInt128 Lower)` tuple field instead. In C# tuples automatically implement `CompareTo`, equality and `GetHashCode`, so you'd be able to simply forward all those methods to the field instead of reimplementing them from scratch.

Comment (RReverser) [reply crates/bindings-csharp/BSATN.Runtime/BSATN/Int256.cs:38]: Signed types might need a bit more care though, not sure.

File: crates/bindings-csharp/BSATN.Codegen/Utils.cs
Comment (RReverser) [reply crates/bindings-csharp/BSATN.Codegen/Utils.cs:136]: In fact, you don't even need to provide one manually - just use lower-endian field order and then you can use `[SpacetimeDB.Type]` for automatic read/write generation on those types.

Comment (RReverser) [reply crates/bindings-csharp/BSATN.Codegen/Utils.cs:136]: I'd just really like to get rid of this hardcoded `switch` in the future - e.g. one idea I had was just defining a `SystemShims` namespace that would be used for any types from the `System` namespace - and adding non-`System` types kinda interferes with that.

I can remove it when I get to it, but I'd prefer not to have non-System types in here to begin with.

File: crates/bindings-csharp/BSATN.Codegen/Utils.cs
Comment (RReverser) [inline crates/bindings-csharp/BSATN.Codegen/Utils.cs:135]: Both need to be checked here, or it will break existing `System.Int128` support in C# modules.
```suggestion
                "System.Int128" or "SpacetimeDB.Int128" => "SpacetimeDB.BSATN.I128",
                "System.UInt128" or "SpacetimeDB.UInt128" => "SpacetimeDB.BSATN.U128",
```

File: crates/bindings-csharp/BSATN.Runtime/BSATN/I128.cs
Comment (RReverser) [inline crates/bindings-csharp/BSATN.Runtime/BSATN/I128.cs:16]: Nit: broken formatting, probably after changing the namespace.

Unfortunately https://github.com/clockworklabs/SpacetimeDB/pull/1567 is still waiting for review, but can you meanwhile install CSharpier and run `dotnet csharpier .` in `crates/bindings-csharp`? That way we have less code to fix once CI check lands.

File: crates/bindings-csharp/Runtime/Internal/Module.cs
Comment (RReverser) [inline crates/bindings-csharp/Runtime/Internal/Module.cs:46]: This shouldn't be necessary as module runtime only targets latest .NET. Let's keep native Int128 here instead rather than the shim.

File: crates/bindings-csharp/Runtime/Runtime.csproj
Comment (RReverser) [inline crates/bindings-csharp/Runtime/Runtime.csproj:14]: This is already enabled for all projects in `Directory.Build.props`, not necessary here.

File: crates/bindings-csharp/BSATN.Runtime/BSATN/Int256.cs
Comment (lcodes) [inline crates/bindings-csharp/BSATN.Runtime/BSATN/Int256.cs]: Autoformat missed the inside of preprocessor conditionals?

File: crates/bindings-csharp/BSATN.Runtime/BSATN/Runtime.cs
Comment (lcodes) [inline crates/bindings-csharp/BSATN.Runtime/BSATN/Runtime.cs]: What happens now if devs declare a field as System.Int128?

File: crates/bindings-csharp/BSATN.Codegen/Utils.cs
Comment (lcodes) [inline crates/bindings-csharp/BSATN.Codegen/Utils.cs]: System.I128?
System.U128?

File: crates/bindings-csharp/BSATN.Runtime/BSATN/Runtime.cs
Comment (lcodes) [inline crates/bindings-csharp/BSATN.Runtime/BSATN/Runtime.cs:178]: Why not call this one U128 and the other UInt128?
Or make both explicit by naming the other U128Sys?

**Review by lcodes — APPROVED** [review]

**Pull request merged**
